mesonbuild/environment.py

```diff
@@ -560,15 +563,61 @@ def __init__(self, source_dir, build_dir, options):
         binaries.build = BinaryTable()
         properties.build = Properties()
 
+        # meson base options
+        _base_options = {}  # type: T.Dict[str, object]
+
+        # Per language compiler arguments
+        compiler_options = PerMachineDefaultable()  # type: PerMachineDefaultable[T.Dict[str, T.Dict[str, object]]]
+        compiler_options.build = {}
+
         ## Read in native file(s) to override build machine configuration
 
+        def load_options(tag: str, store: T.Dict[str, T.Any]) -> None:
+            for section in config.keys():
+                if section.endswith(tag):
+                    if ':' in section:
+                        project = section.split(':')[0]
+                    else:
+                        project = ''
+                    store[project] = config.get(section, {})
+
+        def split_base_options(mopts: T.Dict[str, T.Dict[str, object]]) -> None:
+            for k, v in list(mopts.get('', {}).items()):
+                if k in base_options:
+                    _base_options[k] = v
+                    del mopts[k]
+
+        lang_prefixes = tuple('{}_'.format(l) for l in all_languages)
+        def split_compiler_options(mopts: T.Dict[str, T.Dict[str, object]], machine: MachineChoice) -> None:
+            for k, v in list(mopts.get('', {}).items()):
+                if k.startswith(lang_prefixes):
+                    lang, key = k.split('_', 1)
+                    if compiler_options[machine] is None:
+                        compiler_options[machine] = {}
+                    if lang not in compiler_options[machine]:
+                        compiler_options[machine][lang] = {}
+                    compiler_options[machine][lang][key] = v
+                    del mopts[''][k]
+
         if self.coredata.config_files is not None:
             config = MesonConfigFile.from_config_parser(
                 coredata.load_configs(self.coredata.config_files))
             binaries.build = BinaryTable(config.get('binaries', {}))
-            paths.build = Directories(**config.get('paths', {}))
             properties.build = Properties(config.get('properties', {}))
 
+            # Don't run this if there are any cross files, we don't want to use
+            # the native values if we're doing a cross build
+            if not self.coredata.cross_files:
+                load_options('project options', user_options)
```

Comment:
What exactly is this? Is it something we could do with that defaulting logic instead?

Reply:
This allows you to put project options (ie,` -Duse_systemd`) in your machine file. If we're doing a "native" build, we load those values out of the native file, if we're doing a cross build, load them from the cross file. We don't store them on a per-subproject basis because we only consider the file for the type of build we're doing.